iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -0,0 +1,317 @@
+  auto iteratorTypes = contractionOp.iterator_types().getValue();
+  if (iteratorTypes.size() != 3) {
+    return false;
+  }
+  SmallVector<int, 3> parallel_iterators;
+  SmallVector<int, 3> reduction_iterators;
+  for (int i = 0; i < 3; i++) {
+    if (isParallelIterator(iteratorTypes[i])) {
+      parallel_iterators.push_back(i);
+    } else if (isReductionIterator(iteratorTypes[i])) {
+      reduction_iterators.push_back(i);
+    } else {
+      return false;
+    }
+  }
+  if (parallel_iterators.size() != 2 || reduction_iterators.size() != 1) {
+    return false;
+  }
+  // Give the found iterators some idiomatic names.
+  const int M = parallel_iterators[0];
+  const int N = parallel_iterators[1];
+  const int K = reduction_iterators[0];
+  // Check that there are 3 indexing maps.
+  auto indexingMaps = contractionOp.indexing_maps().getValue();
+  if (indexingMaps.size() != 3) {
+    return false;
+  }
+  // Check that the indexing maps have the expected form.
+  const int expectedMapResults[3][2] = {{M, K}, {N, K}, {M, N}};
+  for (int m = 0; m < 3; ++m) {
+    auto map = indexingMaps[m].cast<AffineMapAttr>().getValue();
+    if (map.getNumDims() != 3 || map.getNumResults() != 2) {
+      return false;
+    }
+    for (int r = 0; r < 2; ++r) {
+      int actualMapResult =
+          map.getResults()[r].cast<AffineDimExpr>().getPosition();
+      if (actualMapResult != expectedMapResults[m][r]) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+// Returns true if `contractionOp` is of the form
+//   matrix * transposed_matrix
+// where matrix is a vector<{mSize}x{kSize}xType>, and
+// transposed_matrix is a vector<{nSize}x{kSize}xType>
+static bool isMatrixTimesMatrixTransposed(vector::ContractionOp contractionOp,
+                                          int64_t mSize, int64_t kSize,
+                                          int64_t nSize) {
+  if (!isMatrixTimesMatrixTransposed(contractionOp)) {
+    return false;
+  }
+  VectorType lhsType = contractionOp.lhs().getType().cast<VectorType>();
+  VectorType rhsType = contractionOp.rhs().getType().cast<VectorType>();
+  auto lhsShape = lhsType.getShape();
+  auto rhsShape = rhsType.getShape();
+  if (lhsShape[0] != mSize || lhsShape[1] != kSize || rhsShape[0] != nSize ||
+      rhsShape[1] != kSize) {
+    return false;
+  }
+  return true;
+}
+
+// Checks that the Value `extResult` is defined by an arith::ExtSIOp promoting
+// from `extSrcType` to `extDstType`, and returns the input of the ExtSIOp.
+static Value getExtSIInput(Type extSrcType, Type extDstType, Value extResult) {
+  auto extSIOp = extResult.getDefiningOp<arith::ExtSIOp>();
+  if (!extSIOp) {
+    return nullptr;
+  }
+  Value extInput = extSIOp.in();
+  if (extInput.getType().cast<VectorType>().getElementType() != extSrcType) {
+    return nullptr;
+  }
+  return extInput;
+}
+
+// Helper to create a 1D, contiguous slice of a 1D vector.
+static Value extract1DSlice(PatternRewriter &rewriter, Location loc,
```

Comment:
Do we even need this method. You could just do

```
rewriter.create<vector::ExtractStridedSliceOp>(
      loc, input, position, dstVecType.getShape(), /*strides=*/1);
```

at the call site.

Reply:
What Geoffrey says.  Regarding the lambda idea, I like that, but the downside would be to have the definition of that (now lambda) function now embedded inside the calling function's body, making it harder to read. I see the pros of both sides but in this specific instance i think the low-tech plain separate functions make for the most readable code.